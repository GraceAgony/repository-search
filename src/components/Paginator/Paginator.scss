.repo-items-list {
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex-wrap: wrap;
  margin-top: 245px;
}

.c-paginator {
  &__loading {
    margin-top: 230px;
  }

  &__empty-result {
    margin-top: 230px;
  }
}
