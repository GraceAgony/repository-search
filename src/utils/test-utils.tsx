import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { createStore, applyMiddleware } from "redux";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import { rootReducer } from "./../redux/store";
import { initialState as reducerInitialState } from "./../redux/modules/repos/reposReducer";

function render(
  ui: JSX.Element,
  {
    initialState = { repositories: reducerInitialState },
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    ...renderOptions
  }: { initialState?: any; store?: any } = {}
) {
  function Wrapper({ children }: any) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from "@testing-library/react";
export { render };
