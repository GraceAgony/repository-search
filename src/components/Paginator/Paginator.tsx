import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import Loader from "react-loader-spinner";
import classnames from "classnames";

import {
  selectReposBySearchString,
  selectMaxPagesNumber,
  selectWereReposLoaded,
  selectSearchString,
} from "redux/modules/repos/reposSelectors";
import { fetchRepos } from "redux/modules/repos/reposActions";

import RepoItem from "components/RepoItem";

import { RootState } from "types";

import "./Paginator.scss";
import "react-loader-spinner/dist/loader/css/react-spinner-loader.css";

interface PaginatorProps {}

const Paginator: React.FunctionComponent<PaginatorProps> = () => {
  const [currentPage, setCurrentPage] = useState(1);
  const dispatch = useDispatch();

  const repositories = useSelector((state: RootState) =>
    selectReposBySearchString(state, currentPage)
  );
  const wereReposLoaded = useSelector(selectWereReposLoaded);
  const currentSearchInput = useSelector(selectSearchString);
  const numberOfPages = useSelector(selectMaxPagesNumber);

  const isPreviousButtonEnabled = currentPage > 1;
  const isNextButtonEnabled = currentPage < numberOfPages;

  const startPage =
    numberOfPages > 4
      ? numberOfPages - currentPage > 2
        ? currentPage - 3 < 0
          ? 1
          : currentPage - 2
        : numberOfPages - 4
      : 1;

  const goToPage = (pageNumber: number) => () => {
    setCurrentPage(pageNumber);
    dispatch(fetchRepos(currentSearchInput, pageNumber));
  };

  const Buttons = () => {
    let buttons = [];
    for (let i: number = 0; i < 5; i++) {
      const currentPageNumber = startPage + i;
      buttons.push(
        <button
          className={classnames("c-paginator__pagination-button", {
            "c-paginator__pagination-button--active":
              currentPageNumber === currentPage,
          })}
          key={currentPageNumber}
          onClick={goToPage(currentPageNumber)}
        >
          {currentPageNumber}
        </button>
      );
    }
    return (
      <div className="c-paginator__pagination-buttons">
        {isPreviousButtonEnabled && (
          <button
            className="c-paginator__pagination-button"
            onClick={goToPage(currentPage - 1)}
          >
            {"<<"}
          </button>
        )}
        {buttons}
        {isNextButtonEnabled && (
          <button
            className="c-paginator__pagination-button"
            onClick={goToPage(currentPage + 1)}
          >
            >>
          </button>
        )}
      </div>
    );
  };

  return wereReposLoaded ? (
    <div className="c-paginator">
      {repositories?.length ? (
        <>
          <div className="repo-items-list">
            {repositories.map((repository) => (
              <RepoItem key={repository.id} {...repository}></RepoItem>
            ))}
          </div>
          <Buttons></Buttons>
        </>
      ) : (
        <div className="c-paginator__empty-result">There are no results</div>
      )}
    </div>
  ) : (
    <div className="c-paginator__loading">
      <Loader type="ThreeDots" color="#0366d6" height={50} width={50} />
    </div>
  );
};

export default Paginator;
