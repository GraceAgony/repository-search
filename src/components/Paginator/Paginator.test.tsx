import React from "react";
import { render, cleanup, screen, fireEvent } from "../../utils/test-utils";
import Paginator from "./Paginator";

const mockedFetch = jest.fn(() => Promise.resolve(""));

beforeEach(() => {
  global.fetch = mockedFetch;
});

afterEach(() => {
  cleanup();
  jest.clearAllMocks();
});

afterAll(() => {
  jest.resetAllMocks();
});

test("should write a snapshot", () => {
  const { container } = render(<Paginator />);
  expect(container).toMatchSnapshot();
});

test("should render correct  number of pages", () => {
  render(<Paginator />, {
    initialState: {
      repositories: {
        error: null,
        loading: false,
        items: { "some search": { totalCount: 68, "1": [{ id: 23232 }] } },
        currentSearchString: "some search",
      },
    },
  });
  expect(screen.queryByText("1")).toBeVisible();
  expect(screen.queryByText("2")).toBeVisible();
  expect(screen.queryByText("3")).toBeVisible();
  expect(screen.queryByText("4")).toBeNull();
});

test("should render correct 1 page if number of results <= 30", () => {
  render(<Paginator />, {
    initialState: {
      repositories: {
        error: null,
        loading: false,
        items: { "some search": { totalCount: 26, "1": [{ id: 23232 }] } },
        currentSearchString: "some search",
      },
    },
  });
  expect(screen.queryByText("1")).toBeVisible();
  expect(screen.queryByText("2")).toBeNull();
});

test("should render correct 5 pagination buttons if number of pages >= 5", () => {
  render(<Paginator />, {
    initialState: {
      repositories: {
        error: null,
        loading: false,
        items: { "some search": { totalCount: 800, "1": [{ id: 23232 }] } },
        currentSearchString: "some search",
      },
    },
  });

  expect(screen.queryByText("5")).toBeVisible();
});

test("should fetch data on pagination button clicked", () => {
  const utils = render(<Paginator />, {
    initialState: {
      repositories: {
        error: null,
        loading: false,
        items: { "some search": { totalCount: 800, "1": [{ id: 23232 }] } },
        currentSearchString: "some search",
      },
    },
  });
  fireEvent.click(utils.getByText("4"));
  expect(mockedFetch).toHaveBeenCalledWith(
    "https://api.github.com/search/repositories?q=some+search&sort=stars&order=desc&page=4&per_page=30"
  );
});
