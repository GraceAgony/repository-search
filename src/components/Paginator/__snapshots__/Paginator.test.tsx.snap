// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should write a snapshot 1`] = `
<div>
  <div
    class="c-paginator"
  >
    <div
      class="c-paginator__empty-result"
    >
      There are no results
    </div>
  </div>
</div>
`;
